DataDog/datadog-agent
# DBM-1824 Add support for emitting resources on metrics via dd.internal.resource tags

### What does this PR do?
Adds support for including resources on metrics emitted by agent integrations. This is to support the [dbm unified instance tagging project](https://docs.google.com/document/d/1onXhg-XVob38C1nHurzLC-s_QqKcmXW4QQdFbMkPuVM). 

This follows the suggestion by @vickenty to use the same approach as `device` and use `resource` tags to "transport" the resources to attach to metrics and then convert the resource tags into actual resources when submitting/serializing to metrics. Any `dd.internal.resource` tag that follows the format `dd.internal.resource:type,value` is converted into a resource and then the tag is removed from the metric tags prior to being sent to metrics. 

The first usage of this is going to be in the `postgres`, `mysql`, `sqlserver` and `oracle` integrations where those integrations will emit metrics with `resource` tags. 

### Motivation

This is part of the [dbm unified instance tagging project](https://docs.google.com/document/d/1onXhg-XVob38C1nHurzLC-s_QqKcmXW4QQdFbMkPuVM). We need to have core database integration metrics to include the associated resources (`database_instance`, `aws_rds_instance`, `gcp_sql_database_instance`, etc) when sending metrics. 

### Additional Notes



### Possible Drawbacks / Trade-offs

The solution you see here minimizes the touch points by using special `resource` tags as the interface for setting resources.  It makes this change easier but it's arguably a bit less visible than taking resources on the C binding aggregator function that's exposed for the python integrations. 

### Describe how to test/QA your changes

The plan is to update one integration (likely postgres) and validate that setting the resource tags are emitted on metrics. The best use case for this that we currently have available to test end-to-end is gcp sql database instances. We can validate that core metrics emitted by the postgres integration for a gcp sql database include the tags set on that resource. 

#### Validation steps ✅ 

In order to validate those changes end-to-end, I did the following:

* I updated our staging pipeline to remove the `database_id` host tag that we currently set there (as a temporary workaround for not having unified instance tagging).
* Using the published docker image from the CI pipeline (`docker.io/datadog/agent-dev:alex-normand-add-support-for-resources-on-metrics-py3-jmx`), I built an agent with [a version of the postgres integration that emitted `gcp_sql_database_instance` resources](https://github.com/DataDog/integrations-core/pull/14185).
* I deployed that agent version to our integration environment monitoring a GCP sql database instance. 

The first step was confirming that the core postgres metrics lost the `database_id` tag prior to the deployment on the agent build described above. Once that was confirmed, the test was to validate that when monitoring the GCP instance with the updated agent build, the core metrics would get tagged with `database_id` (a tag on `gcp_sql_database_instance`) *AND* not have the `dd.internal.resource` tags.

Here's the core postgres metric `postgresql.rows_deleted`:
Without the changes in this PR (but with the pipeline changes to not emit `database_id`):
<img width="912" alt="Screenshot 2023-03-17 at 2 32 03 PM" src="https://user-images.githubusercontent.com/788439/226057477-af6251b5-23ea-4440-b258-d9f7733ea07e.png">

With the agent built from this PR (`dd.internal.resource` is shown to prove that the internal tags aren't sent to metrics):
<img width="946" alt="Screenshot 2023-03-17 at 2 32 18 PM" src="https://user-images.githubusercontent.com/788439/226057518-e9fe2305-1028-4a15-9366-e69563dcb73a.png">

### Reviewer's Checklist
